File: android/app/src/main/java/com/copyeverywhere/app/data/LanDiscoveryDecision.kt
package com.copyeverywhere.app.data

sealed interface LanDiscoveryDecision {
    data object None : LanDiscoveryDecision
    data object WaitForSelection : LanDiscoveryDecision
    data object PreserveManualFallback : LanDiscoveryDecision
    data class AutoSelect(val server: DiscoveredServer) : LanDiscoveryDecision
    data class Restore(val server: DiscoveredServer) : LanDiscoveryDecision
}

object LanDiscoverySelector {
    fun decide(
        servers: List<DiscoveredServer>,
        selectedServer: StoredLanServerSelection?,
        source: LanEndpointSource,
        currentHostUrl: String
    ): LanDiscoveryDecision {
        if (selectedServer != null) {
            val restored = servers.firstOrNull { discovered ->
                !discovered.serverId.isNullOrBlank() && discovered.serverId == selectedServer.serverId
            }
            if (restored != null) {
                return LanDiscoveryDecision.Restore(restored)
            }

            return if (source == LanEndpointSource.ManualFallback) {
                LanDiscoveryDecision.None
            } else {
                LanDiscoveryDecision.PreserveManualFallback
            }
        }

        if (currentHostUrl.trim().trimEnd('/').isNotBlank()) {
            return LanDiscoveryDecision.None
        }

        return when {
            servers.size == 1 -> LanDiscoveryDecision.AutoSelect(servers.first())
            servers.size > 1 -> LanDiscoveryDecision.WaitForSelection
            else -> LanDiscoveryDecision.None
        }
    }
}

File: macos/CopyEverywhere/Sources/CopyEverywhere/LanDiscoveryDecision.swift
import Foundation

enum LanDiscoveryDecision: Equatable {
    case none
    case waitForSelection
    case preserveManualFallback
    case autoSelect(DiscoveredServer)
    case restore(DiscoveredServer)

    static func decide(
        servers: [DiscoveredServer],
        selectedServer: StoredLanServerSelection?,
        source: LanEndpointSource,
        currentHostURL: String
    ) -> LanDiscoveryDecision {
        if let selectedServer {
            if let restored = servers.first(where: { $0.serverID == selectedServer.serverID }) {
                return .restore(restored)
            }

            return source == .manualFallback ? .none : .preserveManualFallback
        }

        let trimmedHostURL = currentHostURL
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard trimmedHostURL.isEmpty else { return .none }

        if servers.count == 1, let server = servers.first {
            return .autoSelect(server)
        }

        if servers.count > 1 {
            return .waitForSelection
        }

        return .none
    }
}
